### PR TITLE
Inlinable and Logic Optimisations

### PR DESCRIPTION
**Pull request by FintanH**
* Add `INLINABLE` pragmas to new `Dhall.Map` functions, as well as `Eq` and `Functor` instances for `Expr`
* Reworked some `Core` normalization logic to just look at the `head` instead of transforming the whole `Map` into a list

### CODE REVIEW

**Comment by FintanH** [comment]
Hey @Gabriel439 I can't seem to reproduce the error in ghci with the counterexample :thinking: Any chance you could give me a hand with it? Thanks!

**Comment by Gabriel439** [comment]
@FintanH: I can reproduce the error:

```
>>> let e = UnionLit "" (Embed (Import {importHashed = ImportHashed {hash = Nothing, importType = Remote (URL {scheme = HTTPS, authority = "", path = File {directory = Directory {components = []}, file = ""}, query = Nothing, fragment = Nothing, headers = Nothing})}, importMode = Code})) (Map (Data.Map.fromList [("",ListIndexed),("a",ListLast)]) ["a",""])
>>> isNormalized e
False
>>> normalize e == e
True
```

It's because of the change you made to the `Eq` instance to `Map`.  I think the right thing to do is to go back to deriving the `Eq` instance for `Dhall.Map.Map` and I'll explain why in the related thread on the other pull request

File: src/Dhall/Core.hs [hunk 1]
```diff
@@ -442,7 +442,144 @@ data Expr s a
     | ImportAlt (Expr s a) (Expr s a)
     -- | > Embed import                             ~  import
     | Embed a
-    deriving (Functor, Foldable, Generic, Traversable, Show, Eq, Data)
+    deriving (Foldable, Generic, Traversable, Show, Data)
+
+instance (Eq s, Eq a) => Eq (Expr s a) where
```
[the remark this answers is not in the document]
Comment (FintanH) [reply src/Dhall/Core.hs]: Good point! I don't think it adds much. I can remove it :)

File: src/Dhall/Map.hs [hunk 2]
```diff
@@ -151,6 +182,33 @@ delete k (Map m ks) = Map m' ks'
...
+{-# INLINABLE lookup #-}
+
+{-| Retrieve the first key/value pair where
+    the key is chosen from the underlying list of keys.
+
+> head k mempty = Nothing
+>
+> head k (singleton k v) = Just (k, v)
+-}
+head :: Ord k => Map k v -> Maybe (k, v)
```
[the remark this answers is not in the document]
Comment (FintanH) [reply src/Dhall/Map.hs]: :100:

File: src/Dhall/Map.hs [hunk 3]
```diff
@@ -230,22 +294,32 @@ mapWithKey :: (k -> a -> b) -> Map k a -> Map k b
...
 
 -- | Traverse all of the key-value pairs in a `Map`, in their original order
 traverseWithKey
     :: Ord k => Applicative f => (k -> a -> f b) -> Map k a -> f (Map k b)
 traverseWithKey f m = fmap fromList (traverse f' (toList m))
   where
     f' (k, a) = fmap ((,) k) (f k a)
+{-# INLINABLE traverseWithKey #-}
+
+traverseWithKey_
```
[the remark this answers is not in the document]
Comment (FintanH) [reply src/Dhall/Map.hs]: Oh interesting! I'll check that out

**Comment by Gabriel439** [comment]
Thank you! 🙂

**Pull request merged**